pkg/ostree/client.js

```diff
@@ -0,0 +1,513 @@
+        self.empty = false;
+
+        promise
+            .done(function(result) {
+                var empty = true;
+                if (transform)
+                    result = transform(result);
+
+                for (var k in result) {
+                    self[k] = result[k];
+                    empty = false;
+                }
+
+                self.empty = empty;
+                self.valid = true;
+            })
+            .fail(function(ex) {
+                self.error = cockpit.message(ex);
+            })
+            .always(function() {
+                self.ready = true;
+                $(self).triggerHandler("changed");
+            });
+    }
+
+    function RPMOSTreeDbusClient() {
+        var self = this;
+
+        self.connection_error = null;
+        self.os_list = [];
+
+        var sysroot = null;
+        var os_proxies = {};
+
+        var os_names = {};
+        var packages_cache = {};
+
+        var local_running = null;
+        var booted_id = null;
+
+        var client = null;
+        var waits = null;
+        var timer = null;
+        var skipped = false;
+
+        Object.defineProperty(this, "running_method", {
+            enumerable: false,
+            get: function() {
+                if (local_running) {
+                    return local_running;
+                } else if (sysroot && sysroot.ActiveTransaction) {
+                    var active = sysroot.ActiveTransaction[0];
+                    var proxy = os_proxies[sysroot.ActiveTransaction[2]];
+
+                    if (proxy && active)
+                        active = active + ":" + proxy.Name;
+
+                    return active;
+                } else {
+                    return null;
+                }
+            }
+        });
+
+        function resolve_nested(obj, path) {
+            return path.split('.').reduce( function( prev, curr ) {
+                if (prev !== undefined)
+                    return prev[curr];
+                else
+                    return prev;
+            }, obj || {} );
+        }
+
+        function trigger_changed() {
+            if (!timer) {
+                $(self).triggerHandler("changed");
+                timer = window.setTimeout(function() {
+                    timer = null;
+                    if (skipped)
+                        $(self).triggerHandler("changed");
+                    skipped = false;
+                }, 300);
+            } else {
+                skipped = true;
+            }
+        }
+
+        function get_client() {
+            if (!client) {
+                var sysroot_ready = false;
+                var os_proxies_ready = false;
+
+                self.connection_error = null;
+                self.os_list = [];
+
+                sysroot = null;
+                os_proxies = {};
+
+                os_names = {};
+                packages_cache = {};
+
+                local_running = null;
+                booted_id = null;
```

Comment:
These lines are all duplicated with above. perhaps a common function would be best?